src/Microsoft.Performance.SDK.Runtime/Extensibility/DataExtensions/ProcessingSystemCompositeCookers.cs

```diff
@@ -0,0 +1,240 @@
+    /// </summary>
+    public sealed class ProcessingSystemCompositeCookers
+          : ICompositeCookerRepository
+    {
+        private readonly ConcurrentDictionary<DataCookerPath, IDataCooker> compositeCookersByPath
+            = new ConcurrentDictionary<DataCookerPath, IDataCooker>();
+
+        private IDataExtensionRepository dataExtensionRepository;
+        private IDataExtensionRetrievalFactory retrievalFactory;
+        private bool disposedValue;
+
+        /// <summary>
+        ///     Initializes an instance of this class.
+        /// </summary>
+        /// <param name="dataExtensionRepository">
+        ///     A data extension repository.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///     <paramref name="dataExtensionRepository"/> is <c>null</c>.
+        /// </exception>
+        public ProcessingSystemCompositeCookers(
+            IDataExtensionRepository dataExtensionRepository)
+        {
+            Guard.NotNull(dataExtensionRepository, nameof(dataExtensionRepository));
+
+            this.dataExtensionRepository = dataExtensionRepository;
+        }
+
+        /// <summary>
+        ///     Initializes this class with a <see cref="IDataExtensionRetrievalFactory"/>.
+        /// </summary>
+        /// <param name="retrievalFactory">
+        ///     Provides access to data needed to instantiate a composite cooker.
+        /// </param>
+        /// <exception cref="ObjectDisposedException">
+        ///     This instance is disposed.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        ///     <paramref name="retrievalFactory"/> is <c>null</c>.
+        /// </exception>
+        public void Initialize(IDataExtensionRetrievalFactory retrievalFactory)
+        {
+            this.ThrowIfDisposed();
+            Guard.NotNull(retrievalFactory, nameof(retrievalFactory));
+
+            this.retrievalFactory = retrievalFactory;
+        }
+
+        /// <inheritdoc/>
+        public ICookedDataRetrieval GetCookerOutput(DataCookerPath dataCookerPath)
+        {
+            this.ThrowIfDisposed();
+
+            Debug.Assert(
+                this.retrievalFactory != null,
+                $"{nameof(Initialize)} needs to be called before accessing composite cookers.");
+
+            return this.GetOrCreateCompositeCooker(
+                dataCookerPath,
+                this.retrievalFactory.CreateDataRetrievalForCompositeDataCooker);
+        }
+
+        internal FilteredCompositeCookers CreateFilteredRepository()
+        {
+            return new FilteredCompositeCookers(this);
+        }
+
+        private ICookedDataRetrieval GetOrCreateCompositeCooker(
+            DataCookerPath cookerPath,
+            Func<DataCookerPath, IDataExtensionRetrieval> createDataRetrieval)
+        {
+            Debug.Assert(createDataRetrieval != null);
+
+            this.ThrowIfDisposed();
+
+            Debug.Assert(
```

Comment:
why not just throw an exception

Reply:
this is a private method, and if we're hitting this, then it's a bug in our code.